text/0000-embroider-v2-package-format.md

```diff
@@ -0,0 +1,817 @@
+
+  The renaming rules allow these addons to adopt V2 format without breaking their public API. New addons _should not_ use renaming rules because it's confusing when the imports people type don't align with their real dependencies.
+
+## Externals
+
+The `externals` property in **Ember package metadata** allows a V2 addon to declare specific imported modules that should not be resolved at build time. Instead they will be resolved at runtime using the traditional `loader.js` `require()`.
+
+When Embroider compiles V1 packages to V2 it does automatic externals detection.
+
+When publishing a native V2 package, any externals need to be listed explicitly in **Ember package metadata**.
+
+An example of when you may need `externals` is when you need to consume a script (not a module) that contains arbitrary `define()` statements. The modules defined by those statements aren't resolvable, in general, at build time. So attempts to import them will generate build errors. By listing them in externals, you can defer them until runtime where they will work.
+
+## Assets
+
+To provide **Assets**, a package includes the `public-assets` key in **Ember package metadata**. It's a mapping from local paths to app-relative URLs that should be available in the final app. For example:
+
+    "name": "my-addon",
+    "ember-addon": {
+      "version": 2,
+      "public-assets": {
+        "./public/image.png": "/my-addon/image.png"
+      }
+    }
+
+with:
+
+    my-addon
+    └── public
+        └── image.png
+
+will result in final build output:
+
+    dist
+    └── my-addon
+        └── image.png
+
+Notice that we’re _not_ choosing to include assets via explicit ECMA `import`. The reason is that fine-grained inclusion of asset files is not critical to runtime performance. Any assets that your app doesn’t actually need, it should never fetch. Assets are always things with their own URLs.
+
+If two V2 packages try to emit assets with the same public URL, that's a build error.
+
+> Q: Should we just automatically namespace them instead?
+> A: That was considered, but it makes backward compatibility harder, and public URLs are not always free to choose/change.
+
+## Build Hooks
+
+In today’s v1 addon packages, the `index.js` file is the main entrypoint that allows an addon to integrate itself with the overall ember-cli build pipeline. The same idea carries forward to v2, with some changes.
+
+It is no longer the `main` entrypoint of the package (see **Own Javascript**). Instead, it’s located via the `build` key in **Ember package metadata**, which should point at a Javascript file. `build` is optional — if you don’t have anything to say, you don’t need the file.
+
+It is now an ECMA module, not a CJS file. The default export is a class that implements your build hooks (there is no required base class).
```

Comment:
What is the reasoning of this decision? Is it using esm for broad node support?

Reply:
It's authored as ES for the same reason that all other code in ember apps and addons is authored as ES. That's the only format that's part of ECMAScript standard.

How it actually runs is an implementation detail (which of course is already true of all the other ES modules in an Ember package).  We could certainly use esm or even babel-to-CJS.